src/logging.c

```diff
@@ -72,6 +103,44 @@ void rcutils_log(
   }
 }
 
+/*!
+  RCUTILS_LOGGING_ENSURE_LARGE_ENOUGH_BUFFER(
+   n, output_buffer_size, allocator, output_buffer, static_output_buffer)
+  Macro for (re)allocating a dynamic buffer if the ouput buffer doesn't have enough space for an
+  additional n characters and null terminator. Whether to allocate or re-allocate is determined by
+  if the output_buffer points to the original static_output_buffer or not. If a new buffer is
+  allocated it will be padded with NUL.
+*/
+
+#define RCUTILS_LOGGING_ENSURE_LARGE_ENOUGH_BUFFER( \
+    n, output_buffer_size, allocator, output_buffer, static_output_buffer) \
+  size_t required_output_buffer_size = strlen(output_buffer) + n + 1; \
+  if (required_output_buffer_size >= output_buffer_size) { \
+    do { \
+      output_buffer_size *= 2; \
+    } while (required_output_buffer_size >= output_buffer_size); \
+    if (output_buffer == static_output_buffer) { \
+      void * dynamic_output_buffer = allocator.allocate(output_buffer_size, allocator.state); \
+      if (!dynamic_output_buffer) { \
+        fprintf(stderr, "failed to allocate buffer for logging output\n"); \
+        goto cleanup; \
+      } \
+      memset(dynamic_output_buffer, '\0', output_buffer_size); \
```

Comment:
I chose to memset the buffer because there are other parts of the code where I'd have to manually copy the null terminator across, and I felt this was safer. I can switch to adding the null terminators manually if we review it extra carefully.

Reply:
Depending on how new data is copied into the buffer I think adding the `\0` there is safer rather than relying on the buffer to be filled with zeros. I don't know if this could happen but if a long message is copied into the buffer first and afterwards a shorter message is copied in (without a resize happening) it must set the trailing `\0` anyway. Therefore I would prefer adding the trailing `\0` whenever a message is written to the buffer. It also keep the logic local and the reader of the code doesn't need to consider in what state the buffer is currently in.

Reply:
alright, I'll switch to that approach and will call attention to that commit for extra-review.

Reply:
3ae0a9b9b473f694252d30701b82d1bb2f971d45 and optionally 1705ae1ead44352790aed2dd701dd0cd67c988b9 are the commits that remove the use of memset in favour of manually copying null terminators into `output_buffer`